OpenDreamShared/Compiler/DM/DMParser.cs

```diff
@@ -8,21 +8,81 @@
 using OpenDreamShared.Compiler.DMPreprocessor;
 
 namespace OpenDreamShared.Compiler.DM {
-    public class DMParser : Parser<Token> {
+    public partial class DMParser : Parser<Token> {
         public static char StringFormatCharacter = (char)0xFF;
 
         private DreamPath _currentPath = DreamPath.Root;
+        private int topLevelErrors = 0;
+        private int topLevelAttempts = 0;
 
         public DMParser(DMLexer lexer) : base(lexer) { }
 
         public DMASTFile File() {
-            DMASTBlockInner blockInner = BlockInner();
+            _saveForToplevel = true;
+            DMASTBlockInner blockInner = TopLevel();
             Newline();
             Consume(TokenType.EndOfFile, "Expected EOF");
+            Console.WriteLine(new String('=', 80));
+            Console.WriteLine("Toplevel statements: " + blockInner.Statements.Length);
+            Console.WriteLine("Errors: " + topLevelErrors + " out of " + topLevelAttempts + " toplevel attempts.");
+            Console.WriteLine(new String('=', 80));
 
             return new DMASTFile(blockInner);
         }
 
+        public DMASTBlockInner TopLevel() {
+            List<DMASTStatement> statements = new();
+            bool valid_state = true;
+            do {
+                Whitespace();
```

Comment:
Why is there a whitespace check at the beginning of a top-level? The lexer should be consuming any whitespace at the beginning of a line, and turning it into indentation.

Reply:
I copied the whitespace logic of Statement() and adapted it to a do/while